File: scripts/showcase.py
#!/usr/bin/env python3
"""
Centurion Theme Showcase - Python
Demonstrates the theme's syntax highlighting with modern Python code patterns.
"""

from dataclasses import dataclass
from enum import Enum
from abc import ABC, abstractmethod


class Priority(Enum):
    """Task priority levels."""

    LOW = 1
    MEDIUM = 2
    HIGH = 3


@dataclass
class Task:
    """A task with title, description, and priority."""

    title: str
    description: str
    priority: Priority = Priority.MEDIUM
    completed: bool = False

    def __repr__(self) -> str:
        status = "✓" if self.completed else "○"
        return f"{status} [{self.priority.name}] {self.title}"


class TaskManager(ABC):
    """Abstract base class for task management."""

    @abstractmethod
    def add_task(self, task: Task) -> None:
        """Add a new task."""
        pass

    @abstractmethod
    def complete_task(self, task_id: int) -> bool:
        """Mark task as completed."""
        pass


class LocalTaskManager(TaskManager):
    """In-memory task manager implementation."""

    def __init__(self) -> None:
        self.tasks: list[Task] = []
        self.completed: int = 0

    def add_task(self, task: Task) -> None:
        """Add task to the list."""
        self.tasks.append(task)
        print(f"✓ Added: {task.title}")

    def complete_task(self, task_id: int) -> bool:
        """Mark task as completed and update counter."""
        if 0 <= task_id < len(self.tasks):
            self.tasks[task_id].completed = True
            self.completed += 1
            return True
        return False

    def get_summary(self) -> dict[str, int]:
        """Get task statistics."""
        high_priority = sum(1 for t in self.tasks if t.priority == Priority.HIGH)
        return {
            "total": len(self.tasks),
            "completed": self.completed,
            "pending": len(self.tasks) - self.completed,
            "high_priority": high_priority,
        }


def create_sample_tasks() -> list[Task]:
    """Generate sample tasks for demonstration."""
    return [
        Task(
            title="Setup development environment",
            description="Install dependencies and configure tools",
            priority=Priority.HIGH,
        ),
        Task(
            title="Implement core features",
            description="Build the main functionality",
            priority=Priority.HIGH,
        ),
        Task(
            title="Write unit tests",
            description="Achieve 85% code coverage",
            priority=Priority.MEDIUM,
        ),
        Task(
            title="Documentation",
            description="Write API documentation",
            priority=Priority.LOW,
        ),
    ]


def main() -> None:
    """Demonstrate the task manager."""
    manager = LocalTaskManager()

    # Add tasks
    for task in create_sample_tasks():
        manager.add_task(task)

    # Complete some tasks
    manager.complete_task(0)
    manager.complete_task(2)

    # Display results
    print("\n" + "=" * 50)
    print("Task Summary")
    print("=" * 50)

    summary = manager.get_summary()
    for key, value in summary.items():
        print(f"{key.replace('_', ' ').title():.<30} {value}")

    print("\n" + "=" * 50)
    print("All Tasks")
    print("=" * 50)

    for idx, task in enumerate(manager.tasks, 1):
        print(f"{idx}. {task}")


if __name__ == "__main__":
    main()

File: scripts/showcase.rs
/**
 * Centurion Theme Showcase - Rust
 * Demonstrates syntax highlighting with modern Rust patterns
 */

use std::fmt;
use std::collections::HashMap;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
enum Priority {
    Low = 1,
    Medium = 2,
    High = 3,
}

impl fmt::Display for Priority {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Priority::Low => write!(f, "LOW"),
            Priority::Medium => write!(f, "MEDIUM"),
            Priority::High => write!(f, "HIGH"),
        }
    }
}

#[derive(Debug, Clone)]
pub struct Task {
    id: u32,
    title: String,
    description: String,
    priority: Priority,
    completed: bool,
}

impl Task {
    pub fn new(id: u32, title: String, description: String, priority: Priority) -> Self {
        Task {
            id,
            title,
            description,
            priority,
            completed: false,
        }
    }

    pub fn complete(&mut self) {
        self.completed = true;
    }

    pub fn is_completed(&self) -> bool {
        self.completed
    }
}

impl fmt::Display for Task {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let status = if self.completed { "✓" } else { "○" };
        write!(
            f,
            "{} [{}] {}",
            status, self.priority, self.title
        )
    }
}

pub struct TaskManager {
    tasks: HashMap<u32, Task>,
    next_id: u32,
}

impl TaskManager {
    pub fn new() -> Self {
        TaskManager {
            tasks: HashMap::new(),
            next_id: 1,
        }
    }

    pub fn add_task(&mut self, title: String, description: String, priority: Priority) -> u32 {
        let id = self.next_id;
        let task = Task::new(id, title.clone(), description, priority);
        self.tasks.insert(id, task);
        self.next_id += 1;
        println!("✓ Added task: {}", title);
        id
    }

    pub fn complete_task(&mut self, id: u32) -> Result<(), String> {
        match self.tasks.get_mut(&id) {
            Some(task) => {
                task.complete();
                Ok(())
            }
            None => Err(format!("Task {} not found", id)),
        }
    }

    pub fn get_task(&self, id: u32) -> Option<&Task> {
        self.tasks.get(&id)
    }

    pub fn get_pending_tasks(&self) -> Vec<&Task> {
        self.tasks
            .values()
            .filter(|task| !task.is_completed())
            .collect()
    }

    pub fn get_high_priority_tasks(&self) -> Vec<&Task> {
        self.tasks
            .values()
            .filter(|task| task.priority == Priority::High && !task.is_completed())
            .collect()
    }

    pub fn get_stats(&self) -> TaskStats {
        let total = self.tasks.len();
        let completed = self.tasks.values().filter(|t| t.is_completed()).count();
        let pending = total - completed;
        let high_priority = self
            .tasks
            .values()
            .filter(|t| t.priority == Priority::High)
            .count();

        TaskStats {
            total,
            completed,
            pending,
            high_priority,
        }
    }
}

impl Default for TaskManager {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug)]
pub struct TaskStats {
    pub total: usize,
    pub completed: usize,
    pub pending: usize,
    pub high_priority: usize,
}

impl fmt::Display for TaskStats {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "Total: {}, Completed: {}, Pending: {}, High Priority: {}",
            self.total, self.completed, self.pending, self.high_priority
        )
    }
}

fn main() {
    let mut manager = TaskManager::new();

    // Add sample tasks
    manager.add_task(
        "Setup development environment".to_string(),
        "Install dependencies and configure tools".to_string(),
        Priority::High,
    );

    manager.add_task(
        "Implement core features".to_string(),
        "Build the main functionality".to_string(),
        Priority::High,
    );

    manager.add_task(
        "Write unit tests".to_string(),
        "Achieve 85% code coverage".to_string(),
        Priority::Medium,
    );

    manager.add_task(
        "Documentation".to_string(),
        "Write API documentation and examples".to_string(),
        Priority::Low,
    );

    // Complete some tasks
    let _ = manager.complete_task(1);
    let _ = manager.complete_task(3);

    // Display statistics
    println!("\n{}", "=".repeat(50));
    println!("Task Statistics");
    println!("{}", "=".repeat(50));
    println!("{}", manager.get_stats());

    // Display pending tasks
    println!("\n{}", "=".repeat(50));
    println!("Pending Tasks");
    println!("{}", "=".repeat(50));
    for task in manager.get_pending_tasks().iter() {
        println!("  {}", task);
    }

    // Display high priority tasks
    println!("\n{}", "=".repeat(50));
    println!("High Priority Tasks");
    println!("{}", "=".repeat(50));
    let high_priority = manager.get_high_priority_tasks();
    if high_priority.is_empty() {
        println!("  No high priority tasks");
    } else {
        for task in high_priority.iter() {
            println!("  {}", task);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_create_task() {
        let task = Task::new(1, "Test".to_string(), "Description".to_string(), Priority::High);
        assert_eq!(task.id, 1);
        assert_eq!(task.priority, Priority::High);
        assert!(!task.is_completed());
    }

    #[test]
    fn test_complete_task() {
        let mut task =
            Task::new(1, "Test".to_string(), "Description".to_string(), Priority::Medium);
        task.complete();
        assert!(task.is_completed());
    }

    #[test]
    fn test_task_manager() {
        let mut manager = TaskManager::new();
        manager.add_task("Task 1".to_string(), "Desc 1".to_string(), Priority::High);
        manager.add_task("Task 2".to_string(), "Desc 2".to_string(), Priority::Low);

        let stats = manager.get_stats();
        assert_eq!(stats.total, 2);
        assert_eq!(stats.completed, 0);
        assert_eq!(stats.pending, 2);
    }
}
